pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -141,15 +228,39 @@ pub mod pallet {
     pub struct BuyCoreCollatorProof<T: Config> {
         account: T::AccountId,
         // TODO
+        // Parth: The signature would be signed using nimbus id. We need to get the author data in this pallet and
+        // verify the collator and its assignment.
+        // Read latest container assignment
         _signature: (),
     }
 
+    /// In flight orders ttl queue
+    #[pallet::storage]
+    pub type InFlightOrdersTtl<T: Config> = StorageValue<
+        _,
+        BoundedBTreeSet<(BlockNumberFor<T>, QueryId), T::MaxInFlightOrders>,
+        ValueQuery,
+    >;
+
     /// Set of parathreads that have already sent an XCM message to buy a core recently.
     /// Used to avoid 2 collators buying a core at the same time, because it is only possible to buy
     /// 1 core in 1 relay block for the same parathread.
     #[pallet::storage]
-    pub type InFlightOrders<T: Config> =
-        StorageValue<_, BoundedBTreeSet<ParaId, T::MaxParathreads>, ValueQuery>;
+    pub type InFlightOrders<T: Config> = StorageValue<
+        _,
+        BoundedBTreeMap<ParaId, InFlightCoreBuyingOrder<BlockNumberFor<T>>, T::MaxInFlightOrders>,
+        ValueQuery,
+    >;
+
+    /// Number of pending blocks
+    #[pallet::storage]
+    pub type PendingBlocks<T: Config> =
+        StorageValue<_, BoundedBTreeSet<ParaId, T::MaxNumberOfParaIds>, ValueQuery>;
+
+    /// Mapping of QueryId to ParaId
+    #[pallet::storage]
+    pub type QueryIdToParaId<T: Config> =
+        StorageValue<_, BoundedBTreeMap<QueryId, ParaId, T::MaxInFlightOrders>, ValueQuery>;
```

Comment:
I've been thinking about these data structures and I think we can do better. It all depends on the max number of requests that we expect to have in flight at one time.

If the number of requests is low (<100), then we don't need `InFlightOrdersTtl` or `QueryIdToParaId`, because we can just iterate over all the para ids in `InFlightOrders` until we find the one we want. The cost of iterating over a list is smaller than serializing a list, and we already serialize it anyway.

If the number of requests is high (>1000), it's complicated, we probably want to use a `StorageMap` instead of a `StorageValue`, but that also has drawbacks. We would need to write some benchmarks.

Reply:
Hmm. I think you may be right. If we have only one data structure that we can iterate, it would reduce the complexity. 

Will implement it.